lib/Sema/CSSolver.cpp

```diff
@@ -1358,6 +1358,221 @@ static Constraint *selectBestBindingDisjunction(
+          auto otherRep =
+              cs.getRepresentative(secondTy->castTo<TypeVariableType>());
+          if (otherRep->isEqual(rep))
+            worklist.push_back(constraint->getFirstType());
+        }
+        break;
+      }
+
+      case ConstraintKind::DynamicTypeOf:
+      case ConstraintKind::EscapableFunctionOf: {
+        auto firstTy = constraint->getFirstType();
+        if (firstTy->is<TypeVariableType>()) {
+          auto otherRep =
+              cs.getRepresentative(firstTy->castTo<TypeVariableType>());
+          if (otherRep->isEqual(rep))
+            worklist.push_back(constraint->getSecondType());
+        }
+        break;
+      }
+
+      case ConstraintKind::OptionalObject: {
+        // Get the underlying object type.
+        auto secondTy = constraint->getSecondType();
+        if (secondTy->is<TypeVariableType>()) {
+          auto otherRep =
+              cs.getRepresentative(secondTy->castTo<TypeVariableType>());
+          if (otherRep->isEqual(rep)) {
+            // See if we can actually determine what the underlying
+            // type is.
+            Type fixedTy;
+            auto firstTy = constraint->getFirstType();
+            if (!firstTy->is<TypeVariableType>()) {
+              fixedTy = firstTy;
+            } else {
+              fixedTy =
+                cs.getFixedType(firstTy->castTo<TypeVariableType>());
+            }
+            if (fixedTy && fixedTy->getOptionalObjectType())
+              worklist.push_back(fixedTy->getOptionalObjectType());
+          }
+        }
+        break;
+      }
+
+      case ConstraintKind::KeyPathApplication:
+      case ConstraintKind::KeyPath: {
+        auto firstTy = constraint->getFirstType();
+        if (firstTy->is<TypeVariableType>()) {
+          auto otherRep =
+              cs.getRepresentative(firstTy->castTo<TypeVariableType>());
+          if (otherRep->isEqual(rep))
+            worklist.push_back(constraint->getThirdType());
+        }
+        break;
+      }
+
+      case ConstraintKind::BindToPointerType:
+      case ConstraintKind::ValueMember:
+      case ConstraintKind::UnresolvedValueMember:
+      case ConstraintKind::Disjunction:
+      case ConstraintKind::CheckedCast:
+      case ConstraintKind::OpenedExistentialOf:
+      case ConstraintKind::ApplicableFunction:
+      case ConstraintKind::BindOverload:
+      case ConstraintKind::FunctionInput:
+      case ConstraintKind::FunctionResult:
+      case ConstraintKind::SelfObjectOfProtocol:
+      case ConstraintKind::ConformsTo:
+        break;
+      }
+    }
+  }
+
+  return false;
+}
+
+// Check to see if we know something about the types of all arguments
+// in the given function type.
+static bool haveTypeInformationForAllArguments(AnyFunctionType *fnType,
+                                               ConstraintSystem &cs) {
+  for (auto param : fnType->getParams())
+    if (!havePotentialTypesOrLiteralConformances(param.getType(), cs))
```

Comment:
Can you use getPlainType() here? getType() wraps the result in an InOutType if its an inout parameter, but is otherwise equivalent. If you need to special-case inouts, just check param.isInOut() instead.

Thanks for using FunctionType::getParams() instead of getInput() though, I appreciate the help in transitioning over to the new representation :-)

Reply:
+1, sorry I keep on forgetting that `getType()` does that... :(